pkg/querybackend/query_profile_entry.go

```diff
@@ -23,39 +24,142 @@ type ProfileEntry struct {
+		fetchPartition: true,
+	}
+	for _, f := range options {
+		if f.iterator != nil {
+			f.iterator(&opts)
+		}
+	}
+	return opts
+}
+
+func profileEntryIterator(q *queryContext, options ...profileIteratorOption) (iter.Iterator[ProfileEntry], error) {
+	opts := iteratorOptsFromOptions(options)
+
+	series, err := getSeries(q.ds.Index(), q.req.matchers, options...)
 	if err != nil {
 		return nil, err
 	}
 	results := parquetquery.NewBinaryJoinIterator(0,
-		q.ds.Profiles().Column(q.ctx, "SeriesIndex", parquetquery.NewMapPredicate(series)),
-		q.ds.Profiles().Column(q.ctx, "TimeNanos", parquetquery.NewIntBetweenPredicate(q.req.startTime, q.req.endTime)),
-	)
-	results = parquetquery.NewBinaryJoinIterator(0, results,
-		q.ds.Profiles().Column(q.ctx, "StacktracePartition", nil),
+		q.ds.Profiles().Column(q.ctx, schemav1.SeriesIndexColumnName, parquetquery.NewMapPredicate(series)),
+		q.ds.Profiles().Column(q.ctx, schemav1.TimeNanosColumnName, parquetquery.NewIntBetweenPredicate(q.req.startTime, q.req.endTime)),
 	)
 
+	columns := []string{
+		schemav1.SeriesIndexColumnName,
+		schemav1.TimeNanosColumnName,
+	}
+	processor := []func([][]parquet.Value, *ProfileEntry){}
+
+	// fetch partition if requested
+	if opts.fetchPartition {
+		results = parquetquery.NewBinaryJoinIterator(0, results,
+			q.ds.Profiles().Column(q.ctx, schemav1.StacktracePartitionColumnName, nil),
+		)
+		offset := len(columns)
+		columns = append(columns, schemav1.StacktracePartitionColumnName)
+		processor = append(processor, func(buf [][]parquet.Value, e *ProfileEntry) {
+			e.Partition = buf[offset][0].Uint64()
+		})
+	}
+	// fetch profile id if requested or part of the predicate
+	if opts.fetchProfileIDs || len(opts.profileIDSelector) > 0 {
+		var predicate parquetquery.Predicate
+		var u uuid.UUID
+		if len(opts.profileIDSelector) > 0 {
+			predicate = parquetquery.NewStringInPredicate(opts.profileIDSelector)
+		}
+		results = parquetquery.NewBinaryJoinIterator(0, results,
+			q.ds.Profiles().Column(q.ctx, schemav1.IDColumnName, predicate),
+		)
+		offset := len(columns)
+		columns = append(columns, schemav1.IDColumnName)
+		processor = append(processor, func(buf [][]parquet.Value, e *ProfileEntry) {
+			b := buf[offset][0].Bytes()
+			if len(b) != 16 {
+				return
+			}
+			copy(u[:], b)
+			e.ID = u.String()
+		})
+	}
+
 	buf := make([][]parquet.Value, 3)
 	entries := iter.NewAsyncBatchIterator[*parquetquery.IteratorResult, ProfileEntry](
 		results, bigBatchSize,
 		func(r *parquetquery.IteratorResult) ProfileEntry {
-			buf = r.Columns(buf,
-				schemav1.SeriesIndexColumnName,
-				schemav1.TimeNanosColumnName,
-				schemav1.StacktracePartitionColumnName)
+			buf = r.Columns(buf, columns...)
 			x := series[buf[0][0].Uint32()]
-			return ProfileEntry{
+			e := ProfileEntry{
 				RowNum:      r.RowNumber[0],
 				Timestamp:   model.TimeFromUnixNano(buf[1][0].Int64()),
 				Fingerprint: x.fingerprint,
 				Labels:      x.labels,
 				Partition:   buf[2][0].Uint64(),
```

Comment:
I think this can fail now. Maybe we cant set it in the processor? This conditionally depends on `opts.fetchPartition`